### ISSUE TITLE
Incorrect syntax highlighting (edge case) on inline macro

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor-intellij-plugin#546]
```
{weight:'dist',default:10,x:'lon',y:'lat'} label:procedure[]
```

![bad1](https://user-images.githubusercontent.com/333276/92233065-e52f6f00-eeaf-11ea-8319-6a00e64f2a21.png)

`label:procedure[]` should be highlighted (inline macro) but `weight` should not.
This is an edge case because, in practice, we will probably use monospaced text on the JSON:

```
`{weight:'dist',default:10,x:'lon',y:'lat'}` label:procedure[]
```

Here's another example:

```
weight:120kg label:procedure[]
```

![weight](https://user-images.githubusercontent.com/333276/92233452-99c99080-eeb0-11ea-856a-6f0b5e0afd61.png)

`weight` is still not an inline macro.

**AsciiDoc Plugin Version**
```
0.30.86
```

**IDE Name and Version**
```
IntelliJ IDEA 2019.3.5 (Ultimate Edition)
Build #IU-193.7288.26, built on May 6, 2020
Licensed to Guillaume GROSSETIE
You have a perpetual fallback license for this version
Subscription is active until November 27, 2020
Runtime version: 11.0.6+8-b520.66 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.15.0-115-generic
GC: ParNew, ConcurrentMarkSweep
Memory: 4029M
Cores: 8
Registry: 
Non-Bundled Plugins: Batch Scripts Support, CMD Support, CheckStyle-IDEA, FindBugs-IDEA, JFormDesigner, com.dmarcotte.handlebars, com.intellij.ideolog, com.jetbrains.lang.ejs, com.jetbrains.plugins.jade, com.wix.eslint, org.asciidoctor.intellij.asciidoc, org.intellij.scala, org.jetbrains.idea.grammar, org.jetbrains.kotlin, org.jetbrains.plugins.slim, org.jetbrains.plugins.ruby
```

**Comment by ahus1** [comment]
As there can be any user-defined macro, the parser is not very restrictive here. Looking at "120kg label:procedure" this could be valid content for an inline macro. Some macros can contain blanks AFAIK (for example xref and image). 

Maybe we could allow blanks in macro names only for well known macros?

Comments would be welcome.

**Comment by mojavelinux** [comment]
Generally it's best to avoid spaces in the target of a macro because it makes it extremely hard to parse (spaces allow for some pretty unlikely scenarios to be matched, so it makes the parser much more aggressive).

**Comment by Mogztter** [comment]
> Looking at "120kg label:procedure" this could be valid content for an inline macro. Some macros can contain blanks AFAIK (for example xref and image).

> Generally it's best to avoid spaces in the target of a macro because it makes it extremely hard to parse (spaces allow for some pretty unlikely scenarios to be matched, so it makes the parser much more aggressive).

TIL: spaces are allowed in the target of a macro 😅

**Comment by mojavelinux** [comment]
> spaces are allowed in the target of a macro

In some cases, but not others. We enabled it for macros that accept a path because people like to put spaces in path names :scream:. But it presents a real challenge for the parser, so a tool could choose to warn against it.

**Comment by Mogztter** [comment]
> In some cases, but not others. We enabled it for macros that accept a path because people like to put spaces in path names 😱.

I see... maybe we could tighten the rule and allow spaces only when the target is surrounded by `++` (similar to `link:++https://example.org/?q=[a b]++`)? Something we could discuss in the spec.

> But it presents a real challenge for the parser, so a tool could choose to warn against it.

I agree 👍 
Though, a warning message might be too much if there's no easy fix. As you said, if the user wants to reference a path (that contains spaces) it's not enough to just remove the spaces in the target.

**Comment by ahus1** [comment]
Spaces work also in `xref`s that reference the title of a section (`xref:This is a Section Heading[]`). I advocate users to use explicit IDs for section headings, but well, current implementations like Asciidoctor support it. 

I'll investigate and see how to implement support for blanks for macros that are "well known" to support blanks.

**Comment by mojavelinux** [comment]
>  As you said, if the user wants to reference a path (that contains spaces) it's not enough to just remove the spaces in the target.

Actually, they can. They can use %20, which was the previous workaround before this "feature" was introduced.

**Comment by mojavelinux** [comment]
> `xref:This is a Section Heading[]`

I strongly recommend against this syntax. If you're using a natural xref (reference by section title, the syntax should be `<<This is a Section Heading>>`. The xref macro is primarily reserved for interdocument xrefs (most commonly used in Antora).

**Comment by ahus1** [comment]
> I strongly recommend against this syntax.

@mojavelinux - me too. Still, the plugin should support this syntax. Future versions of the plugin may issue a warning and offer a refactoring.

**Comment by ahus1** [comment]
The pre-releases 0.30.38 (for IntelliJ 2020.1 and before) and 0.31.24 (for IntelliJ 2020.2 and after) will support the examples @Mogztter provided. 

Implemented logic: Well-known inline macros (including image and xref) that support file names support blanks in the plugin's syntax highlighter and parser. 

To cover some edge cases I decided:

* if they contain blanks in their target, they must not contain colons (`:`) at the same time to avoid most common run-away parsing with incomplete macros and auto-completion.
* auto-completion will not work *after* a blank in a file name if the brackets (`[]`) of the macro are already present. Otherwise it would be too hard to detect where an incomplete macro ends.

![image](https://user-images.githubusercontent.com/3957921/92304793-7c182c00-ef81-11ea-8690-64c1a88283f2.png)

The pre-releases of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://intellij-asciidoc-plugin.ahus1.de/docs/contributors-guide/user/add-eap-repository-to-ide.html).

As I tested with the examples Guillaume provided, I close this issue. Please continue to provide real-live edge cases to help improving syntax highlighting and parsing.

**Comment by Mogztter** [comment]
Thanks @ahus1 it looks great!

**Comment by mojavelinux** [comment]
>  the plugin should support this syntax.

I'll be honest, even I didn't realize that the xref macro supported the natural cross references (cross reference by section title). That was an unintended side effect. And there is no test for it in the test suite.

**Comment by ahus1** [comment]
> I'll be honest, even I didn't realize that the xref macro supported the natural cross references (cross reference by section title). That was an unintended side effect. And there is no test for it in the test suite.

@mojavelinux - thanks for clarifying the preferred syntax. I opened #549 to follow up on this.
